### ISSUE TITLE
Enforced bound TypeVar causes issues with generic ParamSpec

### ISSUE THREAD

**Issue by Bluenix2** [microsoft/pyright#2370]
**Describe the bug**
I have a decorator that returns the same object untouched. In my case I want to also check that this is used on the correct type. PyRight is complaining when I try to return this object.

**To Reproduce**
I have this code:
```python
from typing import Callable, Generic, TypeVar, Union
from typing_extensions import ParamSpec


P = ParamSpec('P')
RT = TypeVar('RT')


class WrappedCallable(Generic[P, RT]):
    callback: Callable[P, RT]

    def __init__(
        self,
        callback: Callable[P, RT],
    ) -> None:
        self.callback = callback

    async def __call__(self, *args: P.args, **kwargs: P.kwargs) -> RT:
        return self.callback(*args, **kwargs)


class AnotherCallable(WrappedCallable[P, RT]):
    ...


T = TypeVar('T', bound='Union[WrappedCallable, AnotherCallable]')


def troublesome_func() -> Callable[[T], T]:
    def decorator(command: T) -> T:
        if not isinstance(command, (WrappedCallable, AnotherCallable)):
            raise ValueError(
                "This decorator can only be applied to WrappedCallable instances"
            )

        ...  # Some code that calls a method on these classes

        return command  # ERROR HERE
    return decorator
```
Which causes the following error:
```
Expression of type "WrappedCallable[Unknown, Unknown]* | AnotherCallable[Unknown, Unknown]*" cannot be assigned to return type "T@troublesome_func"
  Type "WrappedCallable[Unknown, Unknown]* | AnotherCallable[Unknown, Unknown]*" cannot be assigned to type "T@troublesome_func"
```

**Expected behavior**
I have looked into the generics to make sure it isn't my fault but this really looks like it would be correct.

**VS Code extension or command-line**
```
Released on
22/03/2019, 04:10:23
Last updated
04/10/2021, 06:54:57
Identifier
ms-pyright.pyright
```

**Additional context**
If I make the `TypeVar` not bound, or remove the if-statement the errror disappears.

**Comment by gwerbin** [comment]
For what it's worth: I copied and pasted that exact code into `tmp.py`, with the default Pyright settings, and got no error using Pyright v1.1.173:

```
% pyright tmp.py
No configuration file found.
pyproject.toml file found at ...
Loading pyproject.toml file at .../pyproject.toml
Pyproject file ".../pyproject.toml" is missing "[tool.pyright] section.
stubPath .../typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 infos
Completed in 0.762sec
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.177, which I just published. It will also be included in the next release of pylance.
